apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -647,7 +648,7 @@ public async Task WhenSessionPoolReady(CancellationToken cancellationToken = def
             DatabaseName databaseName = Builder.DatabaseName;
             GaxPreconditions.CheckState(databaseName != null, $"{nameof(WhenSessionPoolReady)} cannot be used without a database.");
```

Comment:
This check is now on GetSessionKey, you can remove it from here, and from all the rest.
You also don't need to store the databaseName in a variable now, as it will only be used on GetSessionKey.
And for readability, and for avoiding opening the conneciton unnecessariyl obtain the key first.

Do this for all other methods as well.